kore/src/Kore/Internal/TermLike/TermLike.hs

```diff
@@ -647,34 +648,31 @@ instance HasFreeVariables (TermAttributes variable) variable where
 
 @TermLike@ is the common internal representation of patterns, especially terms.
 
-@TermLike@ is essentially 'Control.Comonad.Cofree.Cofree', but rather than
-define a @newtype@ over @Cofree@, it is defined inline for performance. The
-performance advantage owes to the fact that the instances of 'Recursive.project'
-and 'Recursive.embed' correspond to unwrapping and wrapping the @newtype@,
-respectively, which is free at runtime.
+@TermLike@ is essentially 'Control.Comonad.Cofree.Cofree', but it caches hashes.
 -}
-newtype TermLike variable = TermLike
-    { getTermLike ::
-        CofreeF
-            (TermLikeF variable)
-            (TermAttributes variable)
-            (TermLike variable)
+data TermLike variable = TermLike__
+    { _tlAttributes :: TermAttributes variable -- Should this be strict?
+    -- The hash is stored lazily to avoid computing it unnecessarily. Do we generally
+    -- need the hash? If so, it should probably be strict.
+    , -- | A hash of _tlTermLikeF
```

Comment:
A quick heads-up about this: 
I would definitely prefer if the code was lazy by default but we currently have `Language Strict, StrictData` on by default and turn it of per module. (There is an experimentation ticket to switch to lazy and try to plug holes, to see the performance difference, but always too many things to do).

Reply:
Oof, that's very confusing!